오픈마켓 [Step2] - 산체스

sanches37 opened the pull request:
@GREENOVER 

안녕하세요 그린! Step2 PR을 보냅니다.
step2도 지난프로젝트에서 경험은 해봤지만, 사용하지 않은 기능들을 사용해보려고 노력했습니다.

## 고민한점
* ViewController의 하는 일을 줄이기 위해 collectionView 관련 기능들을 타입으로 분리했습니다. 분리를 하다보니 ViewController에서만 구현할때보다 번거로운 작업들이 많았습니다. 그린은 이런 구현을 어떻게 생각하시는지 궁금합니다.

* 이번 프로젝트에서 리스트형식으로 collectionView를 구현해보기로 했는데, 리스트형식을 tableView가 아닌 collectionView로 구현할 땐 그리드형식을 비롯한 자유로운 형식으로 변환하거나 동시에 쓰기위해 사용할거라 생각해서, 그리드형식도 같이 사용할 수 있도록 구현했습니다.

* 지난 프로젝트에서는 flowLayout을 사용해봐서 compositionalLayout을 사용해봤습니다. 

* 지난 프로젝트에서 NSCache를 적용해봐서 URLCache로 캐시구현을 해봤습니다.  그린은 어떤 방식을 선호하시나요?

* 이미지를 가지고 와서 셀에 적용하기 전에, 셀이 재사용되거나 하는 경우 가지고 온 이미지를 무시하도록 코드를 적용했습니다. 동기 캠퍼들의 코드를 이해하고 참고해서 적용한 부분인데요. cache와 prefetch 를 사용하면서 이러한 경우가 자주 발생하진 않을 것 같은데요. 그린은 제대로 된 동작 구현을 위해 어떤 방법들을 사용하시는지 궁금합니다.

~~~ swift
extension URLSessionProtocol {
    func obtainResponseData(data: Data?, response: URLResponse?, error: Error?) -> Result<Data, NetworkError> {
        let rangeOfSuccessState = 200...299
        if let _ = error {
            return .failure(.dataTaskError)
        }
        guard let response = response as? HTTPURLResponse,
              (rangeOfSuccessState).contains(response.statusCode) else {
            return .failure(.requestFailed)
            
        }
        guard let data = data else {
            return .failure(.dataNotfound)
        }
        
        return .success(data)
    }
}
~~~
* 위와 같이 obtainResponseData을 함수로 처리하고 ImageManager 나 NetworkManager 에서 호출해서 사용하도록 했습니다.
중복되는 코드라 함수처리해서 사용한 게 이유인데요, 코드구현을 해보니 분리로 인해 추가되는 코드들이 생기는 것 같아서 이런 작업을 해주는게 맞는지 의문이 들었습니다.

디코로 항상 응원해주셔서 감사합니다. 🙌😀
바쁘시겠지만, 이번 리뷰도 잘 부탁드립니다. 🙏🏼😆

sanches37 committed:
refactor: step1 관련 수정
제대로 수정되지 않은 valuableHTTPMethod -> applicableHTTPMethod 로 수정
test 메서드에 변경전 네이밍인 ItemsData 와 ItemData 를 사용하는 부분 수정
Media 파일 내 MimeType enum 위치를 RequestEnum.swift 로 이동
NetworkManager 내 if let 을 사용하며 발생한 unused_optional_binding Lint 수정

sanches37 committed:
refactor: APIURL enum 관련 수정
APIURL 연관값 사용하도록 구조 수정
APIURL CustomStringConvertible 채택 해제, 프로퍼티명 path 로 수정
RequestAPI 내 url 프로퍼티 타입 변경
Request 타입내 createRequest 메서드의 매개변수 api 하나만 받도록 수정
ViewController 내 networkManager private 추가

sanches37 committed:
feat: ProductListCell.xib 구현
collectionView 에 사용할 셀 xib 구현

sanches37 committed:
feat: ProductCell 타입 구현
ProductCell 타입 구현하여 productListCell.xib 와 연결

sanches37 committed:
feat: 기본 collectView 구조 구현
UICollectionViewDataSource, UICollectionViewDelegate 채택 후 관련 코드 구현

sanches37 committed:
refactor: URLSession dataTask 수정
URLSession dataTask 부분 공용으로 쓸 수 있도록 메서드 추가
NetworkManager 에 dataTask 공용 메서드 사용
NetworkProtocol.swift -> URLSessionProtocol.swift 로 네이밍 변경

sanches37 committed:
feat: ImageManager 타입 구현
ImageManager 내 loadedImage 메서드 구현
NetworkError 타입 내 convertImageFailed 추

sanches37 committed:
feat: 네트워크 데이터를 불러오는 requestProductList 구현
OpenMarketViewController 내 requestProductList 구현

sanches37 committed:
feat: ProductCell 내 configure 메서드 구현 및 호출
imageConfigure 구현 및 OpenMarketViewController 에서 호출
textConfigure 구현 및 OpenMarketViewController 에서 호출

sanches37 committed:
refactor: Request 내 createRequest 분기 줄이기 위해 수정
boundary uuid 를RequestEnum 내 Boundary enum static 으로 사용하는 구조로 변경

sanches37 committed:
feat: Compositional Layout 적용

sanches37 committed:
refactor: CompositionalLayout 타입으로 분리

sanches37 committed:
feat: ProductGridCell.xib Nib 생성
그리드 collection 에 사용할 ProductGridCell.Xib 생성

sanches37 committed:
refactor: collectionView, ViewController 에서 분리
뷰컨트롤러에서 하는일을 줄이기 위해 collectionView 기능 class 로 분리

sanches37 committed:
feat: segmentedControl 을 이용한 list, grid 화면 구현
segmentCotrol IBOutlet 구현
onCollectionViewTypeChanged IBAction 구현
MarketLayout, Mode 타입 구현
grid 와 list 에 맞게 cell 분기처리
grid 와 list 각각 idetifier 구현

sanches37 committed:
refactor: SegmentedControl 관련 로직 수정
SegmentedControl 관련 로직 수정

sanches37 committed:
feat: strikeThrough 구현 로직 추가
extension String 에 strikeThrough 메서드 구현

sanches37 committed:
feat: Int 타입 세자리 콤마 적용 프로퍼티 구현
extension Int 타입에 세자리 콤마 적용 가능한 withComma 프로퍼티 구현

sanches37 committed:
refactor: ProductCell 로직 수정
awakeFromNib 에 기본 색상 그레이로 지정
strikeThrough, withComma 적용
strikeThrough 메서드에서 프로퍼티로 변경

sanches37 committed:
feat: 셀 재사용시 초기화해줄 prepareForReuse 로직 구현

sanches37 committed:
refactor: 화면 회전시 레이아웃 적용되도록 수정
가로화면시 레이아웃이 적용되도록 구현
앱화면 text 오류로 prepareForReuse 코드 수정

sanches37 committed:
feat: prefetch 적용을 위한 코드 작성
OpenMarketDataSource 타입에 UICollectionViewDataSourcePrefetching 채택
indexPath.item 과 productList.count 를 비교하여 다음페이지 데이터 받게 적용

sanches37 committed:
refactor: 가지고 온 이미지 사용 안될 때 취소 로직 구현
loadedImage 메서드URLSessionTask 를 반환하도록 수정
ProductCell 타입 내 PrepareForReuse 에서 사용 안된 이미지 취소하도록 수정
ProductCell 타입 내 비컨텐츠 PrepareForReuse 에서 리셋하지 않도록 코드 수정
기본으로 사용될 그레이색상 스토리보드에서 적
awakeFromNib 에서 호출되던 항목들 함수처리

sanches37 committed:
refactor: 빠진 super.prepareForReuse() 추가

sanches37 committed:
chore: ProductCell 로 폴더명 변경

sanches37 committed:
feat: URLCache 구현
ImageManager 에 URLCache 관련 코드 추가
AppDelegate 에 제한용량 설정

GREENOVER commented on OpenMarket/OpenMarket/CollectionView/CompositionalLayout.swift:
```diff
@@ -0,0 +1,41 @@
+
+struct CompositionalLayout {
+    
+    enum ScrollDirection {
```
enum 같은 경우 산체스께서 구조화를 잘 해놓으신 Enumeration 폴더에 그룹핑을 하면 어떨까요?
그런다면 유지보수 관리가 더 쉬워지고 여러 구현체가 나눠져서 구조화에 더 좋을것 같습니다🧐

sanches37 replied:
Enumeration 폴더를 만들어 놓고 제대로 사용을 못하고 있었네요. 수정했습니다.

GREENOVER commented on OpenMarket/OpenMarket/CollectionView/CompositionalLayout.swift:
```diff
@@ -0,0 +1,41 @@
+    }
+    
+    func create(portraitHorizontalNumber: Int, landscapeHorizontalNumber: Int, verticalSize: CGFloat, scrollDirection: ScrollDirection) -> UICollectionViewLayout {
+        var horizontalNumber = portraitHorizontalNumber
```
horizontalNumber 변수를 computed property(연산 프로퍼티)로 구현해볼 수 있을것 같은데 산체스 생각은 어떠신가요?
computed property로 구현한다면 현재 저장 프로퍼티로 지정하고 아래에서 다시 조건을 판별하여 저장해주는 코드를 줄일 수 있을거라 생각해요🧐

sanches37 replied:
항상 연산프로퍼티를 잘 안쓰게 되는데요. 말씀해주신대로 수정해서 더 쉽게 보기 좋은 코드가 되었습니다. 감사합니다 😊

GREENOVER commented on OpenMarket/OpenMarket/CollectionView/OpenMarketDataSource.swift:
```diff
@@ -0,0 +1,80 @@
+
+import UIKit
+
+protocol LoadingIndicatable: AnyObject {
```
이건 스타일이겠지만?! 해당 프로토콜이 OpenMarketVC에서 extension해서 사용되고 추후 다른곳에서도 사용될 여지가 있다면 따로 분리해서 관리해도 좋을것 같습니다🧐
나중에 이러한 공통 프로토콜이 많아질때 관리에 용이해요👀
이미 그룹핑 지은 Protocol 폴더가 있네요🤩

sanches37 replied:
이것도 프로토콜 폴더로 이동하겠습니다! 😄

GREENOVER commented on OpenMarket/OpenMarket/CollectionView/OpenMarketDataSource.swift:
```diff
@@ -0,0 +1,80 @@
+                }
+                self.productList.append(contentsOf: product.items)
+                self.nextPage += 1
+                DispatchQueue.main.async {
```
메인 디스패치큐의 역할을 파악하고 적절한 활용이 돋보이네요!👍🏻👍🏻👍🏻

sanches37 replied:
😀🙏🏼

GREENOVER commented on OpenMarket/OpenMarket/CollectionView/CompositionalLayout.swift:
```diff
@@ -0,0 +1,41 @@
+            item.contentInsets = NSDirectionalEdgeInsets(top: 4, leading: 6, bottom: 4, trailing: 6)
+            let groupSize = NSCollectionLayoutSize(widthDimension: .fractionalWidth(1.0), heightDimension: .absolute(verticalSize))
+            
+            if UIDevice.current.orientation.isPortrait {
```
가로 세로에 따른 처리까지.. 크.. 🎉🎉

sanches37 replied:
감사합니다 🙌

GREENOVER commented on OpenMarket/OpenMarket/CollectionView/OpenMarketDelegate.swift:
```diff
@@ -0,0 +1,12 @@
+
+import UIKit
+
+class OpenMarketDelegate: NSObject, UICollectionViewDelegate {
```
요 파일은 구현이 없는데 필요에 의해 살려두신건가요❓
어떤 용도를 가지고 있나요~?

sanches37 replied:
아직은 사용하지 않지만 다음 스텝에서 UICollectionViewDelegate 를 이용하여 셀을 선택하는 기능을 추가하기 위해 만들어뒀습니다.😊

GREENOVER commented on OpenMarket/OpenMarket/CollectionView/OpenMarketDataSource.swift:
```diff
@@ -0,0 +1,80 @@
+        switch sender.selectedSegmentIndex {
+        case 0:
+            changeIdentifier = ProductCell.listIdentifier
+            collectionView.collectionViewLayout = compositionalLayout.create(portraitHorizontalNumber: 1, landscapeHorizontalNumber: 1, verticalSize: 100, scrollDirection: .vertical)
```
요기서 verticalSize는 create에 전달되서 어떤 역할을 하는지 궁금해요😁

sanches37 replied:
셀의 높이를 결정짓는 역할이었습니다. 네이밍이 부족해보여 cellVerticalSize 로 변경했고, 지금은 absolute 사이즈만 사용하게 되어있는데 기능을 선택하여 치수를 줄 수 있도록 수정했습니다. margin 도 매개변수로 전달 받을 수 있게 수정했습니다. 😆

GREENOVER commented on OpenMarket/OpenMarket/CollectionView/OpenMarketDataSource.swift:
```diff
@@ -0,0 +1,80 @@
+    }
+}
+
+extension OpenMarketDataSource: UICollectionViewDataSourcePrefetching {
```
적절한 extension처리 좋네요!!🤩

sanches37 replied:
😀🙏🏼

GREENOVER commented on OpenMarket/OpenMarket/CollectionView/OpenMarketDataSource.swift:
```diff
@@ -0,0 +1,80 @@
+    func stopAnimating()
+}
+
+class OpenMarketDataSource: NSObject {
```
만약에 추가로 테이블뷰를 사용하던지?해서 DataSource라는 네이밍이 필요하다면 OpenMarketDataSource라는 네이밍이 중첩되서 둘다 어떻게 바꿔야할지 고민할 수도 있을것같아요🧐
해당 클래스는 컬렉션뷰의 데이터소스라는 느낌이 있는데 합당한 네이밍은 어떨까요?

sanches37 replied:
말씀하신대로 collectionView 를 명시하기 위해 네이밍 수정을 했습니다.☺️

GREENOVER commented on OpenMarket/OpenMarket/Controller/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,49 @@
+    private let compositionalLayout = CompositionalLayout()
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
```
상위 메서드를 호출하지 않아도 동작하는데 산체스는 이유가 있으니 호출을 해줬겠죠😀
어떤 이유인지 궁금합니다🤩

sanches37 replied:
부모클래스가 완료할 설정을 완료하고 현재 viewDidLoad 를 실행하려는 것인데, 문제 없이 작동할 수도 있지만 경우에 따라서 부모클래스의 설정을 호출해주지 않아 문제가 될 수 있어서  override func 은 습관적으로 super를 호출하고 있습니다 🙂

GREENOVER commented on OpenMarket/OpenMarket/Controller/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,49 @@
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        
+        collectionView.dataSource = openMarketDataSource
```
viewDidLoad 메서드에서 많은 역할을 하고 있는것 같아요!
물론 잘못되었다는건 아닙니다ㅎㅎ
다만 이러한 컬렉션뷰를 구성해주는 역할을 메서드로 나눠서 호출할 수 있을것 같아요😀

sanches37 replied:
메서드로 분리하여 호출하는 방식으로 수정했습니다 😄

GREENOVER commented on OpenMarket/OpenMarket/Controller/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,49 @@
+    }
+}
+
+extension OpenMarketViewController: LoadingIndicatable {
```
이제 프로토콜은 자유자재시군요👍🏻

sanches37 replied:
아닙니다 😅

GREENOVER commented on OpenMarket/OpenMarket/Extension/Data+extension.swift:
```diff
@@ -5,7 +5,7 @@
 //
 
-import Foundation
+import UIKit
```
Foundation에서 UIKit 임포트로 변경하신 이유가 있을까요❓
요거랑 같은 맥락인것 같은데 RequestAPI도 UIKit으로 임포트한 이유가 궁금합니다😀

sanches37 replied:
왜 UIKit 으로 변경했었는지 모르겠네요 다시 Foundation 으로 변경했습니다. 🙂

GREENOVER commented on OpenMarket/OpenMarket/Extension/Int+extension.swift:
```diff
@@ -0,0 +1,17 @@
+
+import Foundation
+
+extension Int {
```
필요한 통화형식을 기본 자료형을 확장해서 사용하는것 아주 최고👍🏻

sanches37 replied:
감사합니다 💪🏻

GREENOVER commented on OpenMarket/OpenMarket/Manager/ImageManager.swift:
```diff
@@ -0,0 +1,51 @@
+        self.session = session
+    }
+
+    func loadedImage(url: String, compleHandler: @escaping (Result<UIImage, NetworkError>) -> Void) -> URLSessionTask? {
```
로드된 이미지라는 메서드명이 조금 어색한것 같습니다🥲
메서드명에 어떤 역할을 하는지 내포되면 더 좋을것 같아요🙌

sanches37 replied:
fetchImage 로 가져오다는 느낌으로 변경해봤습니다 😉

GREENOVER commented on OpenMarket/OpenMarket/Manager/ImageManager.swift:
```diff
@@ -0,0 +1,51 @@
+import UIKit
+
+struct ImageManager {
+    private let cache = URLCache.shared
```
NSCache대신 URLCache를 택하셔서 사용하신 이유를 간단히 저에게 설명해주실수 있을까요❓🙏🏻

sanches37 replied:
사용하지 않은 기능을 써보려는 것과 이번엔 추가로 디스크 Cache 를 구현해보고 싶었는데 URLCache 는 디스크 캐시 기능도 있기 때문에 선택했습니다. 다음번에는 파일매니저를 이용한 방법도 해보고 싶습니다. 😊

GREENOVER commented on OpenMarket/OpenMarket/Manager/NetworkManager.swift:
```diff
@@ -7,43 +7,35 @@
 import Foundation
 
-class NetworkManager {
+struct NetworkManager {
```
class에서 struct로 변경하셨네요👍🏻
어떤 이점을 가져가셨나요?

sanches37 replied:
상속 기능을 사용하지 않거나 mutating func 을 많이 사용하는 구조가 아닐 때는 성능이 더 좋은 struct를 우선적으로 사용하려고 하는데 ,NetworkManager 에는 class 를 사용하지 않아도 될 것 같아서 변경했습니다. 😊

GREENOVER commented on OpenMarket/OpenMarket/Manager/NetworkManager.swift:
```diff
@@ -7,43 +7,35 @@
          applicableHTTPMethod: [APIHTTPMethod] = APIHTTPMethod.allCases) {
         self.session = session
-        self.valuableHTTPMethod = applicableHTTPMethod
+        self.applicableHTTPMethod = applicableHTTPMethod
```
더 의미있고 명확한 네이밍이 되었네요👍🏻

sanches37 replied:
그린 덕분에 더 좋은 네이밍으로 바꿀 수 있었습니다 👍🏼

GREENOVER commented on OpenMarket/OpenMarket/Protocol/URLSessionProtocol.swift:
```diff
@@ -0,0 +1,34 @@
+import Foundation
+
+protocol URLSessionProtocol {
+    func dataTask(with request: URLRequest,
```
파라미터의 시그니처가 포함되어 길어질때 가독이 떨어질때가 많아요.
이경우는 그렇게 가독이 떨어지지 않지만!? 저 같은 경우는 조금이라도 길어진다 싶으면 이런식으로 한눈에 파악되기 쉽게 코드를 작성하는 편입니다🧐
```swift
func dataTask(
    with request: URLRequest,
    completionHandler: @escaping (Data?, URLResponse?, Error?) -> Void
) -> URLSessionTask
```

sanches37 replied:
잘 지정한건지 모르겠는데 길어지는 부분을 전체적으로 수정했습니다.

GREENOVER commented on OpenMarket/OpenMarket/View/ProductCell/ProductCell.swift:
```diff
@@ -0,0 +1,99 @@
+import UIKit
+
+class ProductCell: UICollectionViewCell {
+    @IBOutlet weak var imageView: UIImageView!
```
만약 한 아이템에서 다양한 이미지뷰를 가져와서 보여줘야한다면 이 네이밍은 살아남기 힘들것 같아요🥲
thumbnailImage? productImage 같은 직관적인 네이밍을 추천드립니다😁

sanches37 replied:
thumbnailImage 로 네이밍 변경했습니다. 감사합니다 😆

GREENOVER commented on OpenMarket/OpenMarket/View/ProductCell/ProductCell.swift:
```diff
@@ -0,0 +1,99 @@
+    
+    private let imageManager = ImageManager()
+    private var imageDataTask: URLSessionTask?
+    private static let maximumStockAmount = 999
```
상수 선언 좋아요!!
이런 상수들로 관리하는게 많아지는 코드일수록 이상적인 코드일까요?

sanches37 replied:
휴먼에러를 조금이라도 줄이기 위해 매직넘버를 쓰지 않으려고 지정을 했는데, 이런 상수가 많아진다고 생각하니 또 관리하기에 불편함이 있지 않을까 하는 생각도 드네요 😲

GREENOVER commented on OpenMarket/OpenMarket/View/ProductCell/ProductCell.swift:
```diff
@@ -0,0 +1,99 @@
+    }
+    
+    private func resetContents() {
+        titleLabel.text = nil
```
텍스트들에 nil을 할당해주는 동작은 같은데 변수만 바뀌고 코드가 중복되는
보일러플레이트 코드 느낌이 나는데 혹시 코드중복을 줄여볼 수 있을까요?🧐

sanches37 replied:
이 부분을 해결하기 위해 for loop 같은 방법을 써야 될까요? 조금 더 고민을 해보겠습니다.

sanches37 replied:
forEach 로 nil값을 지정해주도록 수정했습니다. 감사합니다 😀

GREENOVER commented on OpenMarket/OpenMarket/View/ProductCell/ProductCell.swift:
```diff
@@ -0,0 +1,99 @@
+    }
+    
+    func textConfigure(product: Product) {
+        resetContents()
```
만약 이 메서드를 호출해주지 않으면 어떤 사태가 발생하나요❓

sanches37 replied:
메서드를 호출하지 않으면 셀 재사용으로 인해 원하지 않는 데이터가 셀이 올 수 있을 것 같습니다. 
resetContents의 항목들을 prepareForReuse 에서 모두 처리하려고 했지만, 컨텐츠 항목은 prepareForReuse 에서 리셋하는 걸 추천하지 않아서 resetContents에서 처리했었습니다. 컨텐츠 항목의 기준을 어떻게 볼지 애매한 것 같습니다. 
prepareForReuse 에서 초기화하고 있는 thumbnailImage 도 컨텐츠 항목으로 봐야 될 것 같기도 하구요. 🙂

GREENOVER commented on OpenMarket/OpenMarket/View/ProductCell/ProductCell.swift:
```diff
@@ -0,0 +1,99 @@
+        }
+    }
+    
+    func imageConfigure(product: Product) {
```
OpenMarketDataSource에서 textConfigure / imageConfigure을 따로 불러주는데
이것도 좋지만 결국 상품 정보를 configure해준다고 생각합니다.
해당 ProductCell에서 productConfigure이라는 메서드만 호출해서 그 안에서 textConfigure / imageConfigure 등 초기 세팅해줄 수 있는 메서드들을 부르면 더 구조적인 코드가 될것 같아요🧐

sanches37 replied:
감사합니다. 캡슐화가 더 잘 된 것 같습니다!! 😊

GREENOVER commented on OpenMarket/OpenMarketTests/OpenMarketTests.swift:
```diff
@@ -16,7 +16,7 @@ class OpenMarketTests: XCTestCase {
     }
     
-    func test_Items_에셋파일을_ItemsData_타입으로_디코딩에_성공하면_title은_MacBookPro다() {
+    func test_Items_에셋파일을_ProductCollection_타입으로_디코딩에_성공하면_title은_MacBookPro다() {
```
저번 리뷰에도 언급했던 부분인데 추후 이 플젝을 포폴로 쓰시려한다면 네이밍을 생각해봐야할것 같아요😱

sanches37 replied:
영어에 취약해서 적절한 네이밍을 생각해서 스텝3 PR시에는 반영해봐도 될까요? 😩

GREENOVER commented on OpenMarket/OpenMarket/CollectionView/OpenMarketDataSource.swift:
```diff
@@ -0,0 +1,80 @@
+            collectionView.reloadData()
+        default:
+            changeIdentifier = ProductCell.gridItentifier
+            collectionView.collectionViewLayout = compositionalLayout.create(portraitHorizontalNumber: 2, landscapeHorizontalNumber: 4, verticalSize: 250, scrollDirection: .vertical)
```
verticalSize를 주는 부분이 가로/세로 모드를 했을때 영향을 끼치게되는 부분이 있을까요?
질문을 드렸던것은 리스트뷰를 가로모드로 전환하였을때 왼쪽 여백이 너무 많이 생겨버리는 현상이 있습니다🥲

sanches37 replied:
그리드형식에 적용한 margin 이 그대로 적용되어서 그런 것 같습니다. 리스트와 그리드 각자 margin을 주도록 수정하겠습니다 😊

GREENOVER commented on OpenMarket/OpenMarket/Controller/OpenMarketViewController.swift:
```diff
@@ -0,0 +1,49 @@
+extension OpenMarketViewController: LoadingIndicatable {
+    func startAnimating() {
+        loadingIndicator.startAnimating()
+        loadingIndicator.isHidden = false
```
만약 startAnimating과 stopAnimating의 역할이 인디케이터의 display유무를 단지 false/hidden만 시켜준다면
프로토콜 선언부에 녹일수도 있을것 같다는 생각이 들었어요🧐
인자로 받아서 false/true해주게끔? 물론 지금의 코드가 더 나을수 있습니다! 지금도 충~분히 깔끔하고 좋다고 생각해요.
다만 해당 메서드의 네이밍은 startAnimating이라 애니메이션을 시작한다라는 의미인데 isHidden을 처리해주는건 인디케이터 설정에 관련된 로직같아서 굳이굳이 한 함수는 하나의 기능만 하게한다 라는 모토를 최대한 활용하면 그렇게도 적용될 수 있을것 같다는 말씀을 드립니다🤩
다시 말씀드리지만 어떤게 더 나은지는 없습니다! 지금도 좋아요👍🏻👍🏻👍🏻

sanches37 replied:
전적으로 동의합니다. 메서드명에서도 hidden 의 느낌은 찾아볼 수 없었고, 두 기능을 구분해서 구현하는게 맞는 것 같습니다 🙏🏼👍🏼

GREENOVER commented on OpenMarket/OpenMarket/Extension/String+extension.swift:
```diff
@@ -0,0 +1,16 @@
+extension String {
+    var strikeThrough: NSAttributedString {
+        let attributeString: NSMutableAttributedString = NSMutableAttributedString(string: self)
+        attributeString.addAttribute(NSAttributedString.Key.strikethroughStyle, value: NSUnderlineStyle.single.rawValue, range: NSRange(location: 0, length: attributeString.length))
```
요런 경우도 파라미터 선언부가 길어지니 가독성을 더 높일 수 있을것 같아요👀

sanches37 replied:
감사합니다. 이부분도 길어져서 수정했습니다. 😎

GREENOVER commented on OpenMarket/OpenMarket/Manager/NetworkManager.swift:
```diff
@@ -7,43 +7,35 @@
             return
         }
-        guard valuableHTTPMethod.contains(api.httpMethod) else {
+        guard applicableHTTPMethod.contains(api.httpMethod) else {
```
guard를 택하신 이유가 궁금합니다👀
contains를 통해 조건을 비교하고 데이터를 저장해주거나 다른 동작이 없는것 같아서
if문으로 else를 타지 않아도 간결하게 나타낼 수 있지 않을까요🙋🏻‍♂️

sanches37 replied:
많은 생각을 안하고 작성했던 것 같습니다.
if !applicableHTTPMethod.contains(api.httpMethod) {
로 else 를 쓰지 않도록 적용했는데 괜찮을까요?

GREENOVER replied:
네 저도 그렇게 했을것 같습니다!👍🏻

GREENOVER reviewed (changes requested):
안녕하세요 산체스!
스텝1도 엄청 빠르고 잘하셨는데 스텝2는 더 빠르고 잘해주셔서 멋집니다👍👍 
특히 일주일만에 눈에 띄게 코드 가독성도 좋아지고 모듈간 분리에도 신경을 쓰셔서 적용하신 부분이 보여 리뷰하기 편했습니다!
로지컬 띵킹이 무엇인지 확실히 보여주었던 PR인것 같아요🥳
이번 스텝에서는 대부분 간결하고 정확하게 구현해주신것같아 리뷰 코멘트만 읽어보면서 산체스의 의견을 나누고
넘어가도 좋다고 판단되지만 DM으로도 전달드렸듯 필수구현?인 상품등록 화면으로 이동되는 부분 구현이 누락되있어
Request Changes로 요청드리겠습니다😁
아..! 계속 말씀드리지만 리뷰 코멘트는 수정하라는 메시지가 아닌 의견 공유하는 메시지로 봐주시면 감사하겠습니다ㅎㅎ
저도 산체스의 코드에서 많이 배우면서 모르는 부분들을 질문하는 코멘트도 있으니 편하게 같이 소통하면 좋겠습니다👏

***
### 이번에도 칭찬을 드리고 싶은 부분이 아주아주 많아요👏
- 저번 스텝에서 보완해서 네이밍이 훨씬 직관적이고 누구든 이해할 수 있도록 개선되었어요💯
- 프로토콜의 사용을 자유자재로 하시는 부분에서 많이 배워갑니다💯
- 필수 구현사항은 아니였던 부분인것 같은데 사용자를 고려해 가로/세로 모드에 따라 구현하신점💯
- 적절한 extension 처리를 통해 구현부를 분리하고 가독성도 좋아진점💯
- 리스트뿐만 아니라 그리드까지 같이 진행한점 아주 어썸합니다💯

### 궁금한점❓
- 이번에 타이틀로 걸만한 궁금한점은 크게 없었습니다..! (워낙 코드가 술술 읽힌 덕분이겠죠!?)
그래도 자잘한 궁금증들은 리뷰 코멘트로 달아놨으니 의견 나누면 좋을것 같아요🧐
- compositionalLayout를 어떤 이점으로 택했는지 궁금해요ㅎㅎ
좀 더 커스터마이징하는데 용이하다? 그런 이유인지 다른 핵심적인 이유가 있는지 알고싶습니다!

### 산체스의 고민에 대한 저의 생각🤔
- VC에서 저도 작업하다보면 기능이 많아지고 하는 부분에서 산체스와 동일한 고민을 했었고 여전히 많이 하고 있습니다.
제 생각은 지금 산체스가 VC마다 역할을 분리해 파일로 나누고 하는 관리에 대해서 철저히 동의합니다.
제가 생각하는 좋은 구현은 높은 응집도와 낮은 결합성입니다. 지금 산체스의 분리된 코드는 낮은 결합성을 가져가는데 좋아보입니다!
어떻게 하면 응집도는 높이고 결합성은 낮추는 좋은 설계를 어떻게 하는지 사례들이 많아서 한번 이부분도 학습해보시면 지금 산체스의 고민을 타파할 수 있지 않을까 싶습니다!

### 산체스의 추가 구현에 도움이 될것 같은 자료📑
- compositionalLayout를 사용하여 혹시 arrow 데코레이션 버튼이나 separator line이 잘 적용되지 않으셨던건지 궁금합니다!
DM으로 이부분 구현이 아직 안되셨다고 한게 생각나서? 저도 compositionalLayout는 사용해보지 않아서 학습해보며 도움이 되어서 산체스에게도 괜찮을것 같은 자료라고 생각들어요ㅎㅎ (아마 서칭하시다 보셨을수도..!?)
https://idmarystar.github.io/2020-06-28/list-in-uicollectionview-post/
https://demian-develop.tistory.com/21
- NSCache와 URLCache에 대해 고민을 해보셨네요!
우선 두가지를 사용하는 용도가 같다면 어떤게 더 이번 프로젝트에서 이점을 가져가는지 파악해봐야할것 같아요.
밑에 고민포인트와도 연관되는것 같은데 일단 메모리캐시를 이용해서 받아온 데이터들을 전부 prefetch로 저장하는것도 괜찮아요!
지금 프로젝트에서는? 근데 만약 정말 정말 like 당근마켓처럼 데이터가 미친듯이 많다면..? 그럴때 메모리캐시에 계속 누적해서 처음부터 저장하고 사용한다면 조금 비효율적이고 성능 저하가 되지 않을까 싶습니다. (그전에 시스템 디스크 공간이 부족하다면 저장된 캐시가 삭제될수도)
아 주제가 조금 샜는데 그래서 다시 돌아와서!
Foundation 프레임워크가 기본적으로 제공하는 두가지 캐싱 API인데 저같은 경우는 (저도 이부분은 많이 빠삭하진 않지만)
네트워크 데이터 관리일때 거의 URLCache를 사용합니다.
URLCache는 인메모리든 디스크캐시든 데이터를 메모리 청크로 할당하지 않아요.
덕분에 메모리 / 디스크 크기를 커스텀하게 정의할 수 있어 더 플렉시블.. 음 단어가.. 아 유연함! 네..! 좀 더 유연하다고 생각합니다.
근데 저도 사실상 확실히 느껴본 경험은 없어 도움이 되지 않은 답변일수도 있겠네요ㅠㅠ
혹시 산체스는 어떤 차이나 이점때문에 URLCache를 써보셨는지 아니면 저번에 NSCache를 써봐서 이번에 URLCache를 써보신건지 궁금합니다🤔
제 대답이 좀 두루뭉실한데 이해가 안가시면 관련 자료도 보내드리겠습니다!
- prefetch에 대해서 저도 prefetch를 애용하긴 하는데 만약 prefetch를 할 자료들이 진짜 너무너무너무 많다 할때는 어떻게 하면 좋을까요?
그럴 경우 페이지네이션이나 디스패치큐로 백그라운드 다운로드 작업을 통해서도 할 수 있을것 같은데 제대로된 동작은 지금도 괜찮습니다. 
다만 어떤 상황에서 어떤 방법을 가져갈까에 대해 고민할 수 있을것 같아요. 지금 상황에서는 prefetch로 충분하다 생각해요😁
산체스는 이 부분에 대해서 다른 방법들은 어떤게 있을까에 대해 서칭해본게 있다면 코멘트나 DM으로 알려주세요🤔
-  ImageManager 나 NetworkManager에서 호출해서 공통 메서드를 호출해서 사용하는거 저는 구조적으로 괜찮다고 생각합니다!
무조건 코드 중복이 아예 없어야 로직적으로 완벽하고 그러지는 않는다고 생각해요. 최대한 결합성을 낮추는걸 지향하는거지 코드를 분리해서 더 많은 코드를 작성하는게 더 비효율적이지 않을까? 성능에 지장이 있지 않을까?에 대한 고민은 정말 개발자 스타일마다 다를거고 회사마다도 다를거라 생각해요.
그래서 정답은 없지만 제 입장에서는 이러한 코드가 정말 컴파일/빌드 타임에서 아주아주 티도안날정도로 영향을 끼치지 안끼치기에 이럴 경우 지금 산체스처럼 분리를 합니다😁
***
제 리뷰가 산체스의 가려운곳을 조금이라도 긁어드렸는지 싶네요ㅎㅎ
진심으로 이번 PR을 리뷰하면서 산체스가 1주 사이에 정말 많이 학습하고 발전했다는게 느껴졌던 PR입니다!
그런 산체스라서 미비한 구현은 누락하셨던걸로 보여요.
그 부분만 채운 다음에 스텝2를 마무리 하고 싶으시면 수정 후 재 PR 요청을 주시면 될것 같아요😁
정말 이번 스텝 코드 최고👏👏👏

sanches37 committed:
refactor: ScrollDirection 타입 Enumeration 폴더로 이동

sanches37 committed:
refactor: horizontalNumber 연산 프로퍼티로 수정
조건을 프로퍼티안에서 지정할 수 있도록 horizontalNumber 변수 수정

sanches37 committed:
refactor: LoadingIndicatable 프로토콜 폴더로 위치 이동

sanches37 committed:
refactor: CompositionalLayout create 메서드 수정
매개변수로 margin 도 전달 받을 수 있도록 구조 수정
margin 메서드 추가
vierticalSize -> cellVierticalSize 로 네이밍 변경

sanches37 committed:
refactor: collectionView 기능을 명시하는 네이밍 수정
OpenMarketCollectionViewDataSource 로 네이밍 수정
OpenMarketCollectionViewDelegate 로 네이밍 수정

sanches37 committed:
refactor: viewDidLoad  역할 분담위해 메서드 구현
processCollectionView, registeredIdetifier, decidedLayout 메서드 구현

sanches37 committed:
refactor: 불필요한 위치 UIKit import 수정
UIKit -> Foundation import 하도록 수정

sanches37 committed:
refactor: loadedImage -> fetchImage 로 네이밍 변경

sanches37 committed:
style: 길어지는 코드 정리

sanches37 committed:
refactor: imageView -> thumbnailImage 로 네이밍 변경

sanches37 committed:
refactor: productConfigure 만 호출하도록 캡슐화
stockLabel color 지정 빠져서 gray 칼라 지정

sanches37 committed:
refactor: 화살표 악세서리 밑 셀 구분선 구현
CALayer+extension 사용하여 구분선 구현
label 이용하여 화살표 구현

sanches37 committed:
refactor: guard 문 if 문으로 변경
NetworkManager 의applicableHTTPMethod 분기문 if 문으로 수정

sanches37 committed:
refactor: LoadingIndicatable 프로토콜 역할 분리
Animate 와 Hidden 을 분리하여 수행하도록 수정

sanches37 committed:
refactor: ProductCell label for 문으로 작업하도록 수정
label의 text, textColor, attributedText 초기화 for 문으로 작업하게 구현
label의 adjustsFontSizeToFitWidth 값도 for 문으로 구현

sanches37 commented:
그린 리뷰 정말 감사합니다. 부족한 부분이 많이 수정됐다고 생각합니다! 😀🙏🏼
VC에 관해서 높은 응집도와 낮은 결합성을 말씀하셨는데 앞으로도 이런 방향으로 작업할 수 있도록 염두해야겠네요. 감사합니다.
그렇게 설계된 사례들도 찾아볼 수 있도록 하겠습니다. 알려주신 자료도 잘 보겠습니다 🙌
compositionalLayout 레이아웃을 사용한 이유는 리스트형식을 구현하면서 listConfiguration 의 기능과 연관해서 사용하면 
arrow 데코레이션 버튼이나 separator line 을 테이블뷰처럼 쉽게 구현할 수 있을 것이라 생각했고, 지난프로젝트를 할 때 flowLayout을 사용해봐서 다른 방식으로 구현해보고 싶은 마음도 있었습니다. compositionalLayout 도 IOS 13 이후부터나 사용할 수 있어서 선택에 망설이긴 했지만 이번 프로젝트에서는 IOS 14 이전까지는 사용해보자는 생각이었습니다. 공유해주신 자료에서 화살표랑 구분선을 넣을 수 있는 기능이 있는데 IOS14 이후에서 사용할 수 있는 자료들이라 사용하지 않고 다른 방법이 있나 고민을 하다가 먼저 PR을 보냈었습니다. 지금 CFLayer 와 일반 label 을 사용하여 형태만 비슷하게 만들었지만, 아쉬운 구현이긴합니다. 그래도 새로운 레이아웃을 배우며 사용했다는데 만족합니다.
NSCache와 URLCache 에 관해선 위에서 말씀드린 것처럼 디스크 캐시를 같이 적용하려고 고민중에 URLCache를 선택했습니다.
다른 고려사항으로는 NSCache 와 FileManager 를 사용하는 거였는데, 자료가 많이 없었고, 구현하는데 시간이 좀 더 걸릴 것 같아서 다음 기회에 구현해 보려고 합니다. 캐시에 관해 설명도 자세히 해주셔서 감사합니다 🥰
prefatch 에 관련해서는 3가지 방법이 있었는데, 큰 차이는 없어서 기회가 될 때 하나씩 구현해보자는 마음으로 선택했습니다.
아래 링크의 자료를 참고해서 구현했습니다.
https://hanulyun.medium.com/swift-uitableview-에서-pagination을-하는-3가지-방법-ba4a024b58e5
꼼꼼히  리뷰해주셔서 다시 한번 감사드립니다 😃

sanches37 committed:
refactor: 함수명 동사형으로 수정
hidden -> isHidden 으로 변경

GREENOVER commented:
고생하셨습니다 산체스!
이번 스텝2는 진짜 잘하셔서 많이 코멘트 달게 없었습니다👍
그리고 무엇보다 코멘트를 남기면 바로 캐치해서 학습하고 반영하시는게 빨라서 놀랐어요☺️
아 상품등록 화면으로 넘어가는 부분은 아직 구현이 안된것 같은데 이부분은 빈 뷰만 보여주는거라 크게 스텝2에서
재요청을 드릴 부분은 아니라 생각됩니다.
어차피 스텝3에서 진행을 연관해서 해야하니 같이 진행해주시면 될것 같아요😀
이번 스텝도 고생 많으셨습니다!
스텝3부터 정말 본격적으로 산체스가 도전해보고 싶은 부분인것 같은데 화이팅입니다!!
아 제가 좀 늦어진다거나 궁금한점 바로 물어보고 싶으실때는 언제든 디코 디엠 찔러주세요🙏🏻
